test/test_backbone_utils.py

```diff
@@ -89,7 +89,17 @@ def _create_feature_extractor(self, *args, **kwargs):
 
     def _get_return_nodes(self, model):
         set_rng_seed(0)
-        exclude_nodes_filter = ["getitem", "floordiv", "size", "chunk"]
+        exclude_nodes_filter = [
+            "getitem",
+            "floordiv",
+            "size",
+            "chunk",
+            "_assert",
+            "eq",
+            "dim",
+            "getattr",
+            "self_attention",
```

Comment:
I think we missed this line in earlier reviews. I was under the impression that things like `getattr`, `eq` and `dim` were handled already and don't need to be filtered explicitly. Another concern is the addition of `self_attention` here. I assume you add it because it returns a tuple but this probably hints we should handle it differently. 

@alexander-soare any thoughts on whether there is a better way to handle this?

@sallysyw Note that because the model is now moved out of the main area, the tests will no longer run for it on CI automatically. We will need to check any modifications manually.

Reply:
@datumbox making sure we are on the same page: I [originally](https://github.com/pytorch/vision/blob/a7a5818027b95b97f4c04c53e1e10759925a4877/test/test_backbone_utils.py#L58-L70) did it this way because of the way that we **randomly** select any 10 nodes as the return nodes. It's only to facilitate the backward pass test, and doesn't reflect the real usage of the API. The plan was to add filtered node names as and when needed. And I guess it was a bit careless to assume that the same seed wouldn't maybe behave differently on another setup, thereby returning another type of invalid node that hasn't been already accounted for in `exclude_nodes_filter`.

I couldn't think of any way to know the output types of nodes in advance. It could be neater, but a bit cumbersome to explicitly specify a set of node names to choose for each model in turn.

But otherwise I'd just forge ahead with this setup. Addressing the specific lines you raised:

For `self_attention` and generally for any tuple of tensors we can do a type check around [here](https://github.com/pytorch/vision/blob/1d4e2aaebeef6cf9d65b8920c17708a4f9c2e42d/test/test_backbone_utils.py#L157). Maybe replace that line with something like
```python
out_agg = 0
for node_out in out.values():
    if isinstance(node_out, Sequence):
        out_agg += sum(o.mean() for o in node_out)
    # handle dict while we are at it?
    elif isinstance(node_out, Dict):
        out_agg += sum(o.mean() for o in node_out.values())
    else:
        # and the assumption here is that the only other alternative at this point could be a Tensor
        out_agg += node_out.mean()
out_agg.backward()
```

For everything else except `getattr` I was able to verify that they do indeed come through from the feature extraction module:

```python 
import torch
from torchvision.models.feature_extraction import get_graph_node_names


class MyModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.param = torch.nn.Parameter(torch.rand(3, 4))

    def forward(self, x):
        x[0]
        x // 2
        x.size()
        torch._assert(x == x, '')
        x.dim()
        self.param


model = MyModule()

train_nodes, eval_nodes = get_graph_node_names(model)
```
Output is:
```
['x', 'getitem', 'floordiv', 'size', 'eq', '_assert', 'dim', 'param']
```

So I don't know why `get_attr` should be there. That's actually the op associated with the FX node, not the name.

I still feel a little icky about the above though. Thoughts?

Reply:
@alexander-soare Thanks for the detailed comment. I think adopting your proposal makes sense here. We should try to end up as close to the expected output as you have on your comment.

@sallysyw We will probably need to restore the legacy model builders in `torchvision.models` momentarily (perhaps just importing the classes from prototype might work without causing circular dependencies) in order to be able to run the unit-test on the CI and confirm everything still works...

Reply:
The weird thing is that if I comment out `self_attention`, one of the nodes' output would be 
![image](https://user-images.githubusercontent.com/34639474/142963171-82fed9a9-8f82-4e97-baa2-423a3d1a7b07.png)

If I comment out `getattr`, this node would be selected:
![image](https://user-images.githubusercontent.com/34639474/142963248-1de5fd43-cfd2-42f6-b99d-3a9590080ce3.png)

Reply:
@sallysyw Hmm looks like assignment to a tuple like `n, c, h, w = x.shape` results in a gettattr node. So yes,, it should be added to the exclusion as well.

```python
import torch
from torchvision.models.feature_extraction import get_graph_node_names


class MyModule(torch.nn.Module):
    def forward(self, x):
        s, h = x.shape


model = MyModule()

train_nodes, eval_nodes = get_graph_node_names(model)
print(train_nodes)
```
output is: `['x', 'getattr', 'getitem', 'getitem_1']`.

Re your first example with the `None` in the tuple can you print the name of the node? Not sure I'll be able to get the same seed as you to check myself. In any case I suppose you can always add more logic to dodge it like:

```python
out_agg = 0
for node_out in out.values():
    if isinstance(node_out, Sequence):
        out_agg += sum(o.mean() for o in node_out if o is not None)
    # handle dict while we are at it?
    elif isinstance(node_out, Dict):
        out_agg += sum(o.mean() for o in node_out.values() if o is not None)
    else:
        # and the assumption here is that the only other alternative at this point could be a Tensor
        out_agg += node_out.mean()
out_agg.backward()
```

It will always be possible that we need to add yet more logic for another corner case, but short of explicitly naming nodes for each model, we're just going to play this game.

Reply:
> Re your first example with the None in the tuple can you print the name of the node?

```
train_return_nodes = ['encoder.layers.encoder_layer_16.add_1', 'encoder.layers.encoder_layer_18.mlp.act', 'encoder.layers.encoder_layer_0.mlp.dropout_1', 'encoder.layers.encoder_layer_11.ln', 'encoder.layers.encoder_layer_22.mlp.linear_1', 'encoder.layers.encoder_layer_21.mlp.act', 'encoder.layers.encoder_layer_17.mlp.dropout_1', 'encoder.layers.encoder_layer_13.self_attention', 'encoder.layers.encoder_layer_21.dropout', 'encoder.layers.encoder_layer_15.mlp.dropout']
eval_return_nodes = ['encoder.layers.encoder_layer_9.self_attention', 'encoder.layers.encoder_layer_22.mlp.linear', 'encoder.layers.encoder_layer_5.mlp.linear', 'encoder.layers.encoder_layer_12.self_attention', 'encoder.layers.encoder_layer_3.ln_1', 'encoder.layers.encoder_layer_10.mlp.dropout', 'encoder.layers.encoder_layer_23.mlp.linear_1', 'encoder.layers.encoder_layer_5.mlp.dropout_1', 'encoder.layers.encoder_layer_13.ln_1', 'encoder.layers.encoder_layer_3.mlp.act']
```

@alexander-soare